internal/transport/transport_test.go

```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 		t.Errorf("Context timed out")
 	}
 }
+
+// Test that in the event of a graceful client transport shutdown
+// ,  i.e., clientTransport.Close(), if the GOAWAY write is not
+// finished within specified time due to network hang, client
+// should still close without waiting for too long.
+func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
+	// Override timer for writing GOAWAY to 0 so that the connection write
+	// always times out. It is equivalent of real network hang when conn
+	// write for goaway doesn't finish in specified deadline
+	origGoAwayLoopyTimeout := goAwayLoopyWriterTimeout
+	goAwayLoopyWriterTimeout = 0
+	defer func() {
+		goAwayLoopyWriterTimeout = origGoAwayLoopyTimeout
+	}()
+
+	// Create a server.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	// errorCh verifies that desired GOAWAY not received by server
+	errorCh := make(chan error, 1)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Launch the server and handle HTTP/2 connections, specifically
+	// focusing on the reception and interpretation of GOAWAY frames
+	// sent from client.
+	go func() {
+		defer close(errorCh)
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("Error while accepting: %v", err)
+		}
+		defer conn.Close()
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(); err != nil {
+			t.Errorf("Error while writing settings %v", err)
+			return
+		}
+
+		frame, err := framer.ReadFrame()
+		if err != nil {
+			return
+		}
+		switch fr := frame.(type) {
+		case *http2.GoAwayFrame:
```

Comment:
Are we guaranteed to get this GOAWAY frame?

In `http2Client.Close`, since the timeout value is set to `0` in this test, the `case <-time.After(goAwayLoopyWriterTimeout)` will fire before `<-t.writerDone`. And in this case, we set `st` and we go ahead and close the underlying `conn`, by calling `t.conn.Close()` on line 1034, and if this happens before the queued GOAWAY frame was written, then the server would never receive the goaway, right? Or am I missing something?

Reply:
@easwars yeah that's correct and here in this test, server will not receive the GOAWAY as we will get the timeout for  loopy. And we just want to test the timeout here, that is loopy writer doesn't wait for long even if there is network hang.

Reply:
I don't understand. You have code in here which returns without an error when you receive a GOAWAY with code `ErrCodeNo`, but you error out when receive a GOAWAY with any other code.

Reply:
Updated [here](https://github.com/grpc/grpc-go/pull/7371#issuecomment-2232984289)